### DISCUSSION TITLE
How to implement in place overloads

### DISCUSSION THREAD

**Discussion by awni** [pybind/pybind11#5037] [General]
If you have a class bound via Pybin11, what's the right way to implemente inplace overloads so that the returned object is the same as the input object?
```python
      .def(
          "__iadd__",
          [](array& a, const array& v) {  
            a += v;
            return a;
          },
          "other"_a)
```

When checking the ids they are different

```python
x = array(1.0)
print(id(x))
x += y
print(id(x)) # different than first id
```

If that's not enough info to understand the issue, let me know i can provide a more complete example!

**Comment by jnooree** [comment]
When c++ lambda does not have return type specified and you return an lvalue, it defaults to return-by-value.

You'd rather explicitly specify the return type instead:

```cpp
.def("__iadd__", [](array& a, const array& v) -> array& {
    a += v;
    return a;
}, "other"_a)
```

**Reply by awni** [reply]
Thanks, that's exactly what I was looking for!

**Reply by jnooree** [reply]
I also tricked for the exact same reason 🤣
